File: dex/app/src/main/java/com/xayah/dex/ParcelableHelper.java
package com.xayah.dex;

import android.os.Parcel;
import android.os.Parcelable;

public class ParcelableHelper {

    public interface ParcelBlock {
        void accept(Parcel parcel);
    }

    public static void unmarshall(byte[] byteArray, ParcelBlock block) {
        Parcel parcel = Parcel.obtain();
        parcel.unmarshall(byteArray, 0, byteArray.length);
        parcel.setDataPosition(0);
        block.accept(parcel);
        parcel.recycle();
    }

    public static byte[] marshall(Parcelable parcelable) {
        Parcel parcel = Parcel.obtain();
        parcel.setDataPosition(0);
        parcelable.writeToParcel(parcel, 0);
        byte[] bytes = parcel.marshall();
        parcel.recycle();
        return bytes;
    }
}

File: source-next/app/src/main/java/com/xayah/databackup/util/ParcelableHelper.kt
package com.xayah.databackup.util

import android.os.Parcel
import android.os.Parcelable

object ParcelableHelper {
    fun ByteArray.unmarshall(block: (parcel: Parcel) -> Unit) {
        val parcel = Parcel.obtain()
        parcel.unmarshall(this, 0, this.size)
        parcel.setDataPosition(0)
        block(parcel)
        parcel.recycle()
    }

    fun Parcelable.marshall(): ByteArray {
        val parcel = Parcel.obtain()
        parcel.setDataPosition(0)
        this.writeToParcel(parcel, 0)
        val bytes = parcel.marshall()
        parcel.recycle()
        return bytes
    }
}
